### ISSUE TITLE
[TF_Java] Enable Control Dependencies to While Loop Body from Outer Graph

### ISSUE THREAD

**Issue by melissagrueter** [tensorflow/tensorflow#26371]
*System information*

* TensorFlow version (you are using): master
* Are you willing to contribute it (Yes/No): yes

*Describe the feature and the current behavior/state.*

We would like to be able to add control dependencies to ops within the body of a while loop from ops in the outer graph. 

Why: we want to be able to update variables in the outer graph from within the while loop body. An important use case for this is training within a while loop. We understand that in Python, training within a loop is done by adding control dependencies within the loop body from the weight/bias update operations in the outer graph. We had hoped to use the same approach in Java.

Current behavior: We aren't able to add control dependencies from anything outside of the body graph. When we do, we get the following error: 

`Exception in thread "main" java.lang.IllegalArgumentException: Node 'node_in_body_graph': Unknown input node '^node_in_outer_graph'`

We believe this error occurs because in Java, the loop body graph is not connected to the outer graph at the time it is constructed / the time that its control dependencies are added — it is only connected to the outer graph once FinishWhile is called at the C++ level. 

*Will this change the current api? How?*

Yes, people will be able to add control dependencies to operations within the loop body from operations in the outer graph.

*Who will benefit with this feature?*

Anyone who wants to train within their Java while loop, or update variables in the outer graph for any other reason. 

*Any Other info.*

**Comment by melissagrueter** [comment]
cc @skye @sjamesr @karllessard

**Comment by karllessard** [comment]
Actually that was one of my concern, I don’t about the low-level details of the while loop functionality, but ideally there should be a relationship between the inner and the outer graph (e.g. the should be able to share the same `Scope`). 

Right now I don’t have a quick solution for this but it worth it to take some time to think about how to achieve this.

**Comment by skye** [comment]
I assume you mean control deps _from_ the outer graph _to_ the inner while loop graph? You're not allowing to have control deps (or any kind of input) from within a loop body to outside the body.

**Comment by melissagrueter** [comment]
@skye, yes, that is what I meant -- I'll update the above to avoid confusion.

**Comment by skye** [comment]
Got it, thanks!

I agree with Karl, we need some notion of graphs nested in other graphs. In the Python API, this is represented via functions (i.e. [FuncGraphs](http://go/gh/tensorflow/tensorflow/blob/master/tensorflow/python/framework/func_graph.py#L126), which can "capture" Tensors from outer graphs). while_v2 takes advantage of this to handle closed-over inputs inside the loop body. As discussed in a different thread, it'll probably be a lot of work to go down this path for the C/++ APIs.

A quicker solution might be to add the concept of an outer graph to Graph (or maybe just a "friend" Graph? Like in C++ :)), and allow Nodes to have inputs from ancestor graphs. We'd still have a different Scope for each Graph. I'm a little wary of this, because in general it's not safe to use inputs from different Graphs -- this would only work because we eventually copy the inner graph into the outer graph. But I don't think it'd be much code.

**Comment by melissagrueter** [comment]
Sounds good, I will give that approach a try! Thanks @skye and @karllessard.

**Comment by skye** [comment]
Keep us update. There are surely tons of complications I'm not foreseeing... :)

**Comment by karllessard** [comment]
Just out of curiosity @melissagrueter, would eager execution be a better fit for the use case you are trying to cover (i.e. using Java native while loops)?

**Comment by melissagrueter** [comment]
@skye will do!

@karllessard -- we'd like to have the option to do this in a static graph. That said, we're definitely keeping an eye on your eager mode PR and are excited to see where eager mode goes in TF Java!

**Comment by melissagrueter** [comment]
Hi @skye, 

I've been able to add control dependencies from the outer graph to the loop body, but I'm seeing some unexpected behavior: the control input operations are executed a nondeterministic number of times. From what I've seen so far, they run either n, n-1, or n+1 times, where n is the number of loop iterations. If there are multiple control dependencies, they are not all necessarily executed the same number of times in a single run.


To illustrate, here's some quick pseudocode of what I'm seeing:
```
 x = 0  
 y = 0

 //control dependencies 
assign x = x + 1
assign y = y + 1

 i = 0 
while (i < 10) {
  i = i + 1 
 }
```

The assigns (assign x and assign y) are control inputs to either the const (1) or the add operation (i + 1) in the while loop body. Either way, I've seen nondeterminism. Running a session with the loop output as a target and fetching x and y, we might get x = 10 and y = 11. In the next run we might get x = 11 and y = 11, or x = 10 and  y = 9. I don't understand why this nondeterminism is happening.

So far, I've been creating enter nodes to enter the control inputs into the while loop frame (entering constant 1 and variable x and the other constant 1 and variable y). Looking at the protobuf for similar code in Python, however, it appears they enter them into the loop frame by creating enter nodes for the variables (x and y), but adding the Identity of Switch::true as a control input to the constants (both 1). We've tried this in C++, and it seems to get rid of the nondeterminism issue. However, we've only tried this with our own implementation of while loop from primitives where we have access to while loop internals. You mentioned above that control dependencies are not allowed from within the loop body to the outer graph, so I'm not sure this approach will be possible. I'm also not sure why this appears to fix the nondeterminism issue.

If you have any further insight into control dependency behavior that would be super helpful!

Thanks, Melissa

**Comment by skye** [comment]
Hi, sorry for the delay. FYI I switched teams so am no longer actively working on TensorFlow, but I'll continue helping you here until I can find a replacement point person.

I'm a little confused -- in your example, are you trying to update x and y 10 times? And just to clarify, x and y are tf.Variables right? Not just regular tensors. In that case the assigns should be in the while loop body:
```
 x = 0  
 y = 0
 i = 0 
while (i < 10) {
  //control dependencies 
  assign x = x + 1
  assign y = y + 1
  i = i + 1 
}
```
I think you're somehow constructing something like this, since you're seeing x and y be updated multiple times. Did you mean to say that the i + 1 is a control input to the assigns, not the other way around?

In general, the Java while loop code should produce the same thing as the Python code. Even small deviations from what the Python code can produce strange results, as you're finding out :)

**Comment by melissagrueter** [comment]
Hi Skye,

I'm sorry to hear you've left TF; I enjoyed our brief time collaborating! :) Thanks so much for all your help, and also for going above and beyond to still reply here and help us find a new point person — I really appreciate it!

To respond to your points, let me start from the code you sent. Sorry for the confusion!
```
x = 0  
y = 0
i = 0 
while (i < 10) {
  assign x = x + 1
  assign y = y + 1
  i = i + 1 // takes assigns as control inputs
}
```
Although this code would work in Python, it doesn't currently work in Java. It breaks because x and y are not in the body graph, and we get the following error:

`Exception in thread "main" java.lang.IllegalStateException: Input 0 ('Variable') for 'Assign' was not previously added to ShapeRefiner. `

I haven't been able to think of a clean way around this. Maybe a “friend graph” idea like what we discussed with regard to control dependencies earlier would be possible, but from what I saw during my exploration of adding control dependencies, it seems that adding inputs from other graphs is not explicitly disallowed anywhere, so trying to allow it could break things in lots of unexpected places that happen to depend on that assumption. Are you aware of why the implementation of the while loop in C++ creates separate graphs? It seems like it this is the core barrier to mirroring Python behavior. 



I found a relatively simple way to add control inputs from the outer graph to the loop body (by only changing while loop code, not allowing control inputs between separate graphs in general), so I was trying to accomplish the desired behavior this way instead:
```
x = 0  
y = 0
i = 0 
 
assign x = x + 1
assign y = y + 1

while (i < 10) {
  i = i + 1 // takes assigns as control inputs
}
```
This avoids the above issue with x and y since they stay in the outer graph. The dependencies are able to be successfully added due to my changes. However, there's the issue with nondeterminism I mentioned above (from what I've seen, the assigns would nondeterministically be run 9, 10, or 11 times  in the example above). 

 In the spirit of matching Python, I can see that this is likely not a solution to move forward with. However, my understanding of control dependencies is that each time an op executes, its control dependencies are executed, so I'd still expect any operation that's a control input to _i + 1_ to be executed 10 times by the above loop. If you have any further knowledge of control dependencies that could explain where “each time an op executes, its control dependencies are executed” falls short, that would be super helpful —  I'd love to have a clearer understanding for reference on any future work involving control dependencies.

**Comment by skye** [comment]
This is a great explanation, thank you! My perspective has become too Python API-centric apparently...

Can you share the nondeterministic GraphDef? I have some theories about what could be going on, but I think it'll be easier to explain if I can look at the GraphDef to make sure they make sense.

**Comment by skye** [comment]
If it's too big to easily post here, feel free to email it to me.

**Comment by melissagrueter** [comment]
Thanks @skye! Just sent you an email.

**Comment by skye** [comment]
Hi Melissa, here's what I think is happening... since the Variables are entered into the loop with Enter(is_constant=True) nodes, they are available in _all frames_ of the while loop (a frame roughly corresponds to an iteration, and since is_constant=True, the Enter outputs are always available). A new frame is created every time the NextIteration node runs, meaning that on the final iteration, a new frame will be created after the final iteration ends, and then the loop will exit. I believe you're seeing 11 when that last frame, that isn't actually used, causes the Assign to run one extra time. The Python code creates a control dep on the Switch::true output to prevent the Assign from running on this last frame, since the Switch::false branch is taken when the loop exits.

I'm less sure about how you're getting 9, but my guess is that the Assign is running multiple times in parallel (since nodes from different iteration frames can run in parallel as long as there's no pending dependencies), causing some of the updates to be lost. The control dep on the switch would fix this too, since it effectively causes the iterations to run in order (at least for the Assign).

Does this make sense? It barely makes sense to me :) I unfortunately don't have any clever solutions at the moment... I think we might need to change the C API while loop API, this is too hard to work with.

**Comment by melissagrueter** [comment]
Hi @skye,

Thanks so much for taking a look! Your explanation makes sense to me. :) 

As for potential changes to the C API loop implementation, do you think making it take lambdas to build the cond and body in the outer graph instead of creating separate cond/body graphs would work? That way I think we could actually do things the Python way (reproduced below again) since x and y would be in the same graph as the loop body. 

```
x = 0  
y = 0
i = 0 
while (i < 10) {
  assign x = x + 1
  assign y = y + 1
  i = i + 1 // takes assigns as control inputs
}
```

Or were you envisioning other changes to the implementation?

Also, do you know who maintains the C API that we should consult on this? 

Thanks again for all your help!!!

**Comment by skye** [comment]
That's one option. We could also revisit allowing the underlying Graph to allow inputs from Tensors outside the Graph. Like you pointed out, this could lead to a bunch of unexpected problems, but we could at least limit it to only be allowed in the TF_Graphs returned by TF_NewWhile, and the advantage would be avoiding changes to the C API. Lemme see if I can someone still working on TF to weigh in...

**Comment by skye** [comment]
cc @saxenasaurabh 

Based on some offline discussion, I think we're leaning towards adding the ability to "capture" Tensors in a TF_Graph. This is very similar to allowing inputs from other Graphs. This would be a step towards representing functions as graphs in the C API, which is probably a prerequisite for while_v2, so it'd be nice to move towards that final goal.

**Comment by melissagrueter** [comment]
Hi @skye,

After thinking about it some more, I'm still confused about why the Assign runs in the final frame. In general, what triggers control dependencies to run? 

My working assumption had been that dataflow determines the execution flow of the graph, so an op's control dependencies would only be triggered to run when its inputs flow in. 

In the graph I sent you, the Assign was a control input to a constant in the loop body. In this case, the constant's inputs are obviously available (since there are none) and the Assign is also available, so I can see why the Assign might run in the final frame if it happens to be run before the constant's other control dependency on switch::true. (Side question: is my observation that an op's control dependencies are run in an unpredictable order correct?) 

However, I've also tried making the Assign a control input to an op which takes switch::true as an input, and I still see the off-by-one error in that case. There, I would have assumed that because the op's inputs never become available, the op's control dependencies would not be triggered. This must mean an op's control dependencies can be run before the op's inputs become available. 

If not inputs flowing in, what triggers an op's control dependencies to run? Are all the control dependencies in a frame run once per frame, when it's created?

**Comment by skye** [comment]
The assigns run in the final frame (and every frame) because they are loop invariants due to the RefEnters having is_constant=true (since the Variables aren't loop variables). Specifically:

All constant enters are marked as loop invariants:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/executor.cc#L2132
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/executor.cc#L2779
In our case, these are the RefEnters for the Variables.

The constant enters are activated/run for every loop iteration frame:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/executor.cc#L2785 (existing frames)
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/executor.cc#L2820 (new frames added after the Enter runs for the first time)

Running the enters then flows through to any downstream ops, in our case to the Assigns. It doesn't matter to the Assigns that there's a control output to the Const, since an op running depends on only on its upstream inputs, no downstream outputs. So we need a control edge from the Switch::true to the Assigns, so the true branch is upstream of the Assigns. Does this make sense?

**Comment by melissagrueter** [comment]
Hi @skye, 

Ahh, everything makes sense now. :) Thank you so much for all the time you've taken to write such thorough responses to all my questions! I really appreciate all your help.

**Comment by skye** [comment]
No problem! Please let me know if you have more questions.

Would you or someone on your team be willing to implement capturing in TF_Graphs (or some other solution)? Or wait until post-TF 2.0? The TF team is focused on TF 2.0 right now, so realistically there probably isn't internal bandwidth to fix this until after 2.0.

**Comment by melissagrueter** [comment]
Hi @skye,

Yes, I'd like to contribute whatever needed to get the while loop working! Is there anyone you'd recommend trying to pull into the discussion when I open the issue/feature request for it?

**Comment by saxenasaurabh** [comment]
Hi @melissagrueter 

Feel free to pull me into any discussions.

Thanks!

**Comment by melissagrueter** [comment]
Thanks @saxenasaurabh! Will do.
